internal/weaver/routing.go

```diff
@@ -28,22 +28,44 @@ import (
 
 // routingBalancer balances requests according to a routing assignment.
 type routingBalancer struct {
-	balancer  call.Balancer // default balancer
+	balancer  call.Balancer // balancer to use for non-routed calls
 	tlsConfig *tls.Config   // tls config to use; may be nil.
 
 	mu         sync.RWMutex
 	assignment *protos.Assignment
 	index      index
+
+	// Map from address to connection. We currently allow just one
+	// connection per address.
+	// Guarded by mu.
+	conns map[string]call.ReplicaConnection
 }
 
 // newRoutingBalancer returns a new routingBalancer.
 func newRoutingBalancer(tlsConfig *tls.Config) *routingBalancer {
-	return &routingBalancer{balancer: call.RoundRobin(), tlsConfig: tlsConfig}
+	return &routingBalancer{
+		balancer:  call.RoundRobin(),
+		tlsConfig: tlsConfig,
+		conns:     map[string]call.ReplicaConnection{},
+	}
 }
 
-// Update implements the call.Balancer interface.
-func (rb *routingBalancer) Update(endpoints []call.Endpoint) {
-	rb.balancer.Update(endpoints)
+// Add adds c to the set of connections we are balancing across.
+func (rb *routingBalancer) Add(c call.ReplicaConnection) {
+	rb.balancer.Add(c)
```

Comment:
I don't think RoundRobin balancers are thread safe. Should we do this with `rb.mu` held as well?

Reply:
I think this might be okay?

Balancer requires external synchronization. (It is protected by rc.mu in call.go.) The comment in balancer.go isn't great though; I will improve it.

routingBalancer.update() provides a path where rc.mu is not held, but that state is guarded by routingBalancer.mu and does not touch the balancer part.